### PR TITLE
DbLedgerStorage -- Write cache

### PR DESCRIPTION
**Pull request by merlimat**
1st part of DbLedgerStorage related changes.

This PR introduces a `WriteCache` class that is used to store entries before they get flushed on the entryLogs. 

The key part is that it provides a way to iterate in order over the entries and it is is garbage-free at steady state.

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        // Align to 64 bytes so that different threads will not contend the same L1 cache line
+        int alignedSize = align64(size);
+
+        long offset;
+        int localOffset;
+        int segmentIdx;
+
+        while (true) {
+            offset = cacheSize.getAndAdd(alignedSize);
+            localOffset = (int) (offset & SegmentOffsetMask);
```
Comment (ivankelly) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: The mixture of & and / are a bit weird here. & requires that SegmentOffsetMask be a power of 2, minus 1. This being the case, then getting the segment index should be a bitwise shift, not a division.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+        assertArrayEquals(expectedSorted, data);
+    }
+
+    @Test
+    public void keySmallerThanTotalSize() {
+        try {
+            new ArrayGroupSort(3, 2);
+            fail("should have failed");
+        } catch (IllegalArgumentException e) {
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java]: Nit: This can be put in @Test(expected=IllegalArgumentException.class)

Same for below checks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        sortedEntriesLock.lock();
+
+        try {
+            int entriesToSort = (int) index.size();
+            int arrayLen = entriesToSort * 4;
+            if (sortedEntries == null || sortedEntries.length < arrayLen) {
+                sortedEntries = new long[(int) (arrayLen * 2)];
+            }
+
+            long startTime = System.nanoTime();
```
Comment (ivankelly) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: Use MathUtils here and in the calculations further down.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+            startTime = System.nanoTime();
+
+            // Sort entries by (ledgerId, entryId) maintaining the 4 items groups
+            groupSorter.sort(sortedEntries, 0, sortedEntriesIdx);
+            if (log.isDebugEnabled()) {
+                log.debug("sorting {} ms", (System.nanoTime() - startTime) / 1e6);
+            }
+            startTime = System.nanoTime();
+
+            ByteBuf[] entrySegments = new ByteBuf[segmentsCount];
```
Comment (ivankelly) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: i don't understand why you're making this copy here.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        if (lastEntryId == -1) {
+            // Ledger not found in write cache
+            return null;
+        } else {
+            return get(ledgerId, lastEntryId);
+        }
+    }
+
+    public void deleteLedger(long ledgerId) {
+        deletedLedgers.add(ledgerId);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: We can delete a ledger during a foreach. This is handled gracefully using concurrent collections.
But isn't it bad to have a foreach to return only part of the entries in case of a delete happening during the scan?
Maybe it will not be possible I am just asking about this case just to understand

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        try {
+            sorter.sort(new long[] { 1, 2 });
+            fail("should have failed");
+        } catch (IllegalArgumentException e) {
+            // ok
+        }
+    }
+
+    @Test
+    public void emtpy() {
```
Comment (ivankelly) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java]: typo: empty

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCacheTest.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        assertTrue(cache.isEmpty());
+        assertEquals(0, cache.count());
+        assertEquals(0, cache.size());
+
+        entry1.release();
+        cache.close();
+    }
+
+    @Test
+    public void cacheFull() throws Exception {
```
Comment (ivankelly) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCacheTest.java]: Could do with a few more tests.
- ledger deletion when there's more than one ledger.
- Edge cases on buffer/segment sizes
- Multiple threads accessing at the same time.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java (hunk 1 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: Very good point. I absolutely cannot remember why is this not a `>> 32`. Need to think about it.

Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: Actually, I think the result is the same. 

The logic is that the offset keeps increasing when adding each entry. 
If each segment is 2Gb (maxInt), and the current offset is 2.5Gb, the `segmentIdx` will be 1 after the division. The 31 bits shift will be faster though.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java (hunk 4 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: This is making a copy of the cache buffers for each segment (just the ByteBuf shell objects) so that we can play with the `readerIdx` and `writerIdx` and pass on always the same `ByteBuf` instance to the iteration consumer. See later: 

```java
ByteBuf entry = entrySegments[segmentIdx];
entry.setIndex(localOffset, localOffset + (int) length);
consumer.accept(ledgerId, entryId, entry);
```

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+public class WriteCache implements Closeable {
+
+    /**
+     * Consumer that is used to scan the entire write cache.
+     */
+    public interface EntryConsumer {
+        void accept(long ledgerId, long entryId, ByteBuf entry);
+    }
+
+    private final ConcurrentLongLongPairHashMap index = new ConcurrentLongLongPairHashMap(4096,
```
[the remark this answers is not in the document]
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: Fixed

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+            2 * Runtime.getRuntime().availableProcessors());
+
+    private final ConcurrentLongLongHashMap lastEntryMap = new ConcurrentLongLongHashMap(4096,
+            2 * Runtime.getRuntime().availableProcessors());
+
+    private final ByteBuf[] cacheSegments;
+    private final int segmentsCount;
+
+    private static final int MaxSegmentSize = Integer.MAX_VALUE;
```
[the remark this answers is not in the document]
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: Fixed

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java (hunk 3 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: Sure

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java (hunk 6 above)
Comment (merlimat) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java]: Fixed

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java (hunk 2 above)
Comment (merlimat) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/ArraySortGroupTest.java]: Fixed

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCacheTest.java (hunk 7 above)
Comment (merlimat) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCacheTest.java]: Done

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java (hunk 5 above)
Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/WriteCache.java]: If the ledger is deleted during a foreach, some entries for that ledger may be flushed and some may not. That is handled anyway when the entryLogs are compacted.

**Comment by sijie** [comment]
@eolivelli : @merlimat addressed your comments. please take a look again.

**Comment by sijie** [comment]
@eolivelli sorry I missed your review :(

**Comment by sijie** [comment]
@merlimat seems there are a few checkstyle errors
